c10/util/Exception.cpp

```diff
@@ -147,25 +147,44 @@ thread_local WarningHandler* ThreadWarningHandler::warning_handler_ = nullptr;
 
 } // namespace
 
-void warn(
+Warning::Warning(
     const SourceLocation& source_location,
     const std::string& msg,
-    const bool verbatim) {
-  ThreadWarningHandler::get_handler()->process(source_location, msg, verbatim);
-}
+    const bool verbatim)
+    : source_location_(std::move(source_location)),
+      msg_(std::move(msg)),
```

Comment:
You cannot move from a `const&`. If you want to move from it, you should pass these things by value. This is sometimes referred to as the "copy-and-move idiom".